### PR TITLE
Ну вот и всё

### PR DESCRIPTION
**Pull request by Lobochkin**
---
:mortar_board: [Ну вот и всё](https://up.htmlacademy.ru/yii/1/user/248841/tasks/23)

### CODE REVIEW

File: frontend/controllers/AccountController.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    public function actionIndex()
+    {
+
+        $model = new Account();
+
+        if (\Yii::$app->request->getIsPost() && \Yii::$app->request->isAjax) {
+            $model->load(\Yii::$app->request->post());
+            if ($model->validate()) {
+                if (boolval($_FILES) && !$model->saveFiles()) {
+                    return 'Ошибка сохранения файлов!';
```
Comment (denis-khodakovskiy) [inline frontend/controllers/AccountController.php]: Exception

File: frontend/controllers/MyTasksController.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                        'status' => \Lobochkin\TaskForce\Task::STATUS_CANCEL,
+                        'status' => \Lobochkin\TaskForce\Task::STATUS_FAILED,
+                    ])
+                    ->orderBy('id DESC')
+                    ->orWhere([
+                        'implementer_id' => \Yii::$app->user->getId(),
+                        'employer_id' => \Yii::$app->user->getId(),
+                    ]);
+                break;
+            case 'delay':
```
Comment (denis-khodakovskiy) [inline frontend/controllers/MyTasksController.php]: const

File: frontend/models/Account.php [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+
+    /**
+     * @return bool
+     * @throws \yii\base\Exception
+     * @throws \yii\db\Exception
+     */
+    public function saveFiles()
+    {
+        $transaction = \Yii::$app->db->beginTransaction();
+        $userId = \Yii::$app->user->identity->getId();
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: $this->user->id

File: frontend/models/Account.php [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+
+        if ($_FILES['Account']['tmp_name']['avatarUrl'] || $_FILES['avatarUrl']['tmp_name']) {
+            $user = Users::findOne($userId);
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: $this->user

File: frontend/models/Account.php [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+     * @return bool
+     * @throws \yii\base\Exception
+     * @throws \yii\db\Exception
+     */
+    public function saveFiles()
+    {
+        $transaction = \Yii::$app->db->beginTransaction();
+        $userId = \Yii::$app->user->identity->getId();
+
+        if ($_FILES['Account']['tmp_name']['avatarUrl'] || $_FILES['avatarUrl']['tmp_name']) {
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: if ($avatarUrl = UploadedFile::getInstanceByName('Account[avatarUrl]')) {...

File: frontend/models/Account.php [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+            $user = Users::findOne($userId);
+
+            if ($user->avatar_url) {
+                unlink(\Yii::getAlias('@webroot/uploads/') . $user->avatar_url);
+            }
+
+            $avatarUrl = UploadedFile::getInstanceByName('Account[avatarUrl]');
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: drop

File: frontend/models/Account.php [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+
+            $user->avatar_url = $filenameRandom;
+
+            if (!$user->save()) {
+                $transaction->rollBack();
+                return false;
+            }
+        }
+
+        if ($_FILES['files']) {
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: if ($files = UploadedFile::getInstancesByName('files')) {...

File: frontend/models/Account.php [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+            $files = UploadedFile::getInstancesByName('files');
+            if (!$files) {
+                return false;
+            }
+            if (count($files) > 6) {
+                $files = array_slice($files, 0, 6);
+            }
+
+            $photos = PhotoWork::find()->where(['user_id' => $userId])->all();
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: $this->user->photoWork

File: frontend/models/Account.php [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+    public function saveAccount()
+    {
+        $transaction = \Yii::$app->db->beginTransaction();
+        $this->user->name = $this->name;
+        $this->user->city_id = $this->cityId;
+        $this->user->email = $this->email;
+        $this->user->birthday = (new \DateTime($this->birthday))->format('Y-m-d');
+        $this->user->description = $this->description;
+        $arCat = $this->user->getUserCategory()->select(['category_id', 'active'])->asArray()->all();
+        $newCategory = !$this->category ? [] : $this->category;
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: $fromDB = $this->user->userCategory;
        $fromRequest = $this->category;

        $toActivateOrAdd = array_diff($fromRequest, $fromDB);
        $toDeactivate = array_diff($fromDB, $fromRequest);

        foreach ($toDeactivate as $categoryId) {
            //..
        }

        foreach ($toActivateOrAdd as $categoryId) {
            $affectedRowsCount = UserCategory::updateAll([
                'active' => 'Y',
            ], [
                'user_id' => $this->user->id,
                'category_id' => $categoryId,
            ]);

            if (!$affectedRowsCount) {
                $model = new UserCategory();
                $model->user_id = $this->user->id;
                $model->category_id = $categoryId;
                $model->active = 'Y';
                $model->save();
            }
        }

File: frontend/models/Account.php [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+        }
+
+        if ($this->password1){
+            if ($this->password1 !== $this->password2){
+                $transaction->rollBack();
+                return 'Пароли не совпадают';
+            }
+            $this->user->password = \Yii::$app->security->generatePasswordHash($this->password1);
+        }
+        $this->user->phone = substr($this->phone,0,12);
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: drop,
preg_replace ^\d

File: frontend/models/Account.php [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+        $this->user->skype = $this->skype;
+        $this->user->telegram = $this->telegram;
+
+
+        if ($this->notShowProfile === '1'){
+            $this->user->not_show_profile = 'Y';
```
Comment (denis-khodakovskiy) [inline frontend/models/Account.php]: `$this->user->not_show_profile = $this->notShowProfile === 1 ? 'Y' : 'N';`

**Review by denis-khodakovskiy — CHANGES_REQUESTED** [review]

File: frontend/controllers/MyTasksController.php (hunk 2 above)
Comment (Lobochkin) [reply frontend/controllers/MyTasksController.php]: сделал

File: frontend/models/Account.php (hunk 3 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 4 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 5 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 7 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 8 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 10 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 11 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/models/Account.php (hunk 9 above)
Comment (Lobochkin) [reply frontend/models/Account.php]: сделал

File: frontend/controllers/AccountController.php (hunk 1 above)
Comment (Lobochkin) [reply frontend/controllers/AccountController.php]: сделал

**Pull request merged**